### PR TITLE
PIO support for RP2040

### PR DESCRIPTION
**Pull request by mateir-7**
### Pull Request Overview

This pull request adds PIO support for RP2040.

### Testing Strategy

This pull request was tested by running PIO examples on a Pico Explorer Base that can be found at the end of the main.rs file.

### TODO or Help Wanted

No help needed.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -6,4 +6,5 @@
         "RUSTUP_TOOLCHAIN": "nightly-2024-07-08"
     },
     "rust-analyzer.check.allTargets": false,
+    "makefile.configureOnOpen": false,
```
Comment (bradjc) [inline .vscode/settings.json:9]: Can we move this to a different PR?

File: chips/rp2040/src/pio.rs [hunk 2]
```diff
@@ -0,0 +1,1231 @@
...
+#[derive(Clone, Copy)]
+pub enum PioMovStatusType {
+    StatusTxLessthan = 0,
+    StatusRxLessthan = 1,
+}
+
+// PIO State Machine configuration structure
+//
+// Used to initialize a PIO with all of its state machines.
+pub struct StateMachineConfiguration {
```
Comment (bradjc) [inline chips/rp2040/src/pio.rs:587]: All of these struct comments should be `///`.

File: chips/rp2040/src/pio.rs [hunk 3]
```diff
@@ -0,0 +1,1231 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright OxidOS Automotive 2024.
+//
+// Author: Radu Matei <radu.matei.05.21@gmail.com>
+
+use kernel::debug;
```
Comment (bradjc) [inline chips/rp2040/src/pio.rs:7]:
```suggestion
// Licensed under the Apache License, Version 2.0 or the MIT License.
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright OxidOS Automotive 2024.
//
// Author: Radu Matei <radu.matei.05.21@gmail.com>

//! Programmable Input Output (PIO) hardware.

use kernel::debug;
```

**Review by bradjc — COMMENTED** [review]
Maybe we can move the changes to main.rs to a test file within the board directory?

File: .vscode/settings.json (hunk 1 above)
Comment (mateir-7) [reply .vscode/settings.json:9]: Yes, sure!

File: chips/rp2040/src/pio.rs (hunk 2 above)
Comment (mateir-7) [reply chips/rp2040/src/pio.rs]: Thank you!

File: chips/rp2040/src/pio.rs (hunk 3 above)
Comment (mateir-7) [reply chips/rp2040/src/pio.rs]: Done, thanks for the help.

**Comment by mateir-7** [comment]
> Maybe we can move the changes to main.rs to a test file within the board directory?

Yes, I left them there hoping that it would maybe help you test it, but then thought that it may be helpful to people who want to use the driver. Do I just make another file in src for the examples or in an entire different directory?

File: chips/rp2040/src/chip.rs [hunk 4]
```diff
@@ -176,6 +181,13 @@ impl<'a> Rp2040DefaultPeripherals<'a> {
 impl InterruptService for Rp2040DefaultPeripherals<'_> {
     unsafe fn service_interrupt(&self, interrupt: u32) -> bool {
         match interrupt {
+            interrupts::PIO0_IRQ_0 => {
+                // As the PIO HIL doesn't provide any support for interrupts, they are
```
Comment (hudson-ayers) [inline chips/rp2040/src/chip.rs:185]: What does PIO HIL refer to here?

File: chips/rp2040/src/pio.rs [hunk 5]
```diff
@@ -0,0 +1,1233 @@
...
+
+use crate::gpio::{GpioFunction, RPGpio, RPGpioPin};
+
+const NUMBER_STATE_MACHINES: usize = 4;
+const NUMBER_INSTR_MEMORY_LOCATIONS: usize = 32;
+
+#[repr(C)]
+struct InstrMem {
+    // Write-only access to instruction memory locations 0-31
+    instr_mem: ReadWrite<u32, INSTR_MEMx::Register>,
```
Comment (hudson-ayers) [inline chips/rp2040/src/pio.rs:22]: Tock registers does have support for a `WriteOnly` type -- seems like maybe that is what you want here given the comment?

File: chips/rp2040/src/pio.rs [hunk 6]
```diff
@@ -0,0 +1,1233 @@
...
+    pub out_pull_threshold: u32,
+    pub jmp_pin: u32,
+    pub out_special_sticky: bool,
+    pub out_special_has_enable_pin: bool,
+    pub out_special_enable_pin_index: u32,
+    pub mov_status_sel: PioMovStatusType,
+    pub mov_status_n: u32,
+    pub div_int: u32,
+    pub div_frac: u32,
+}
```
Comment (hudson-ayers) [inline chips/rp2040/src/pio.rs:615]: I am not too familiar with how PIO is used in practice, but am a bit surprised that you expect all the individual fields of this struct to be accessible outside of this crate.

File: chips/rp2040/src/pio.rs [hunk 7]
```diff
@@ -0,0 +1,1233 @@
...
+            x += 1;
+            if x == 32 {
+                break;
+            }
+        }
+    }
+
+    /// Clears all of a PIO instance's instruction memory.
+    pub fn clear_instr_registers(&self) {
+        for i in 0..31 {
```
Comment (hudson-ayers) [inline chips/rp2040/src/pio.rs:1059]: [Rust ranges are only half open](https://doc.rust-lang.org/std/ops/struct.Range.html), i.e. they exclude the second value from the iteration. I suspect this should be `0..=31` or `0..32`. Even better would be defining a constant representing the number of instructions that fit in memory per PIO instance.

File: chips/rp2040/src/pio.rs [hunk 8]
```diff
@@ -0,0 +1,1233 @@
...
+        self.sm_config(sm_number, config);
+        self.pio_number = pio_number;
+        self.gpio_init(&RPGpioPin::new(RPGpio::from_u32(pin1)));
+        self.gpio_init(&RPGpioPin::new(RPGpio::from_u32(pin2)));
+        self.sm_set_enabled(sm_number, false);
+        self.set_pins_out(sm_number, pin1, 1, true);
+        self.set_pins_out(sm_number, pin2, 1, true);
+        self.sm_init(sm_number);
+        self.sm_set_enabled(sm_number, true);
+        self.sm_put_blocking(sm_number, 0b11000000);
```
Comment (hudson-ayers) [inline chips/rp2040/src/pio.rs:1148]: could you use a named constant that shows what this binary value means?

File: chips/rp2040/src/pio.rs [hunk 9]
```diff
@@ -0,0 +1,1233 @@
...
+        self.sm_config(sm_number, config);
+        self.pio_number = pio_number;
+        self.gpio_init(&RPGpioPin::new(RPGpio::from_u32(pin)));
+        self.sm_set_enabled(sm_number, false);
+        self.set_pins_out(sm_number, pin, 1, true);
+        self.set_side_set_pins(sm_number, pin);
+        self.sm_init(sm_number);
+        self.sm_put_blocking(sm_number, pwm_period);
+        self.sm_exec(sm_number, 0x8080_u32); // pull
+        self.sm_exec(sm_number, 0x60c0_u32); // out isr, 32
```
Comment (hudson-ayers) [inline chips/rp2040/src/pio.rs:1169]: same here -- named constants would be great

**Review by hudson-ayers — COMMENTED** [review]
Thanks for this contribution -- excited to have PIO support in Tock. A few comments.

File: chips/rp2040/src/pio.rs (hunk 7 above)
Comment (mateir-7) [reply chips/rp2040/src/pio.rs:1059]: Thank you, switched it to the constant used for the registers, NUMBER_INSTR_MEMORY_LOCATIONS - also used it in add_program to make sure only 32 instructions are uploaded per program.

File: chips/rp2040/src/pio.rs (hunk 8 above)
Comment (mateir-7) [reply chips/rp2040/src/pio.rs:1148]: Sure, switched this to turn_on_gpio_6_7 and added a comment explaining how it works.

File: chips/rp2040/src/pio.rs (hunk 9 above)
Comment (mateir-7) [reply chips/rp2040/src/pio.rs:1169]: Fixed, same as before, added constants and comments explaining the exact PIO commands to make searching easier.

File: chips/rp2040/src/pio.rs (hunk 6 above)
Comment (mateir-7) [reply chips/rp2040/src/pio.rs:615]: Honestly not sure, was thinking that if an user wanted to make a program in another folder or just in main.rs for example they might need to change just a few fields and leave the others as default - but to use them they need to be pub. Any ideas are welcome, I really don't know if this is the right approach.

File: chips/rp2040/src/chip.rs (hunk 4 above)
Comment (mateir-7) [reply chips/rp2040/src/chip.rs:185]: Completely missed this, thank you. Put this as a placeholder probably at some point and forgot to change it - it is just the PWM comment changed to PIO. IRQs in PIO work as registers - not sure how an interrupt handler would work though, as the user/creator of the program would use them as they wish.

File: chips/rp2040/src/pio.rs [hunk 10]
```diff
@@ -0,0 +1,1238 @@
...
+const PIO1_XOR_BASE: StaticRef<PioRegisters> =
+    unsafe { StaticRef::new((PIO_1_BASE_ADDRESS + 0x1000) as *const PioRegisters) };
+const PIO1_SET_BASE: StaticRef<PioRegisters> =
+    unsafe { StaticRef::new((PIO_1_BASE_ADDRESS + 0x2000) as *const PioRegisters) };
+const PIO1_CLEAR_BASE: StaticRef<PioRegisters> =
+    unsafe { StaticRef::new((PIO_1_BASE_ADDRESS + 0x3000) as *const PioRegisters) };
+
+// There are a total of 4 State Machines per PIO.
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum SMNumber {
```
Comment (bradjc) [inline chips/rp2040/src/pio.rs:503]: These need to all be `///` comments.

**Review by bradjc — CHANGES_REQUESTED** [review]
I think it's important we support the PIO and this is great. Using precompiled byte code seems fine as long as it is contained to the chips/rp2040 directory. I think the last thing to do is implement one or more HIL interfaces to use the PIO as a specific IO controller. We actually do something similar (although there is no bytecode interpreter thing) for the USART in the SAM4L chip (one hardware peripheral can implement multiple HIL interfaces).

File: chips/rp2040/src/pio.rs (hunk 10 above)
Comment (mateir-7) [reply chips/rp2040/src/pio.rs]: Done, hopefully got all of them.

**Comment by alevy** [comment]
I second @bradjc's request that what's remaining here (other than nits) is to replace the example PIO applets in the board `main` with at least one "real" device and an implementation for the relevant HIL. E.g., a PWM driver that uses PIO and implements `kernel::hil::pwm::Pwm`

File: chips/rp2040/src/pio_pwm.rs
Comment (bradjc) [inline chips/rp2040/src/pio_pwm.rs:1]: Need a copyright block and `//!` header comment.

**Review by bradjc — CHANGES_REQUESTED** [review]

File: chips/rp2040/src/pio_pwm.rs
Comment (mateir-7) [reply chips/rp2040/src/pio_pwm.rs:1]: Done!

File: chips/rp2040/src/pio_pwm.rs [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+use kernel::utilities::cells::OptionalCell;
+use kernel::{hil, ErrorCode};
+
+pub struct PioPwm<'a> {
+    clocks: OptionalCell<&'a clocks::Clocks>,
+}
+
+impl<'a> PioPwm<'a> {
+    pub fn start_pwm() {}
+}
```
[the remark this answers is not in the document]
Comment (mateir-7) [reply chips/rp2040/src/pio_pwm.rs:14]: Was WIP, modified

**Comment by mateir-7** [comment]
Done, left out a few lines in main.rs if you want to test the pwm, everything was moved into pio_pwm - tell me if anything needs changing. Need to test with oscilloscope and see if formula for translating frequency from hz to pio cycles is correct. Otherwise this might be done

**Review by bradjc — DISMISSED** [review]

File: chips/rp2040/src/chip.rs (hunk 4 above)
Comment (hudson-ayers) [reply chips/rp2040/src/chip.rs:185]: Either way I think we should remove the reference to a nonexistent HIL. I think it is ok if figuring out how interrupts work for PIO is left for the first PR that actually uses them

File: chips/rp2040/src/pio.rs (hunk 6 above)
Comment (hudson-ayers) [reply chips/rp2040/src/pio.rs:638]: Fair, if we expect the state machine logic to live outside this file it might be hard to predict which fields need to be accessible. Fine to leave this as-is

File: chips/rp2040/src/chip.rs (hunk 4 above)
Comment (alevy) [inline chips/rp2040/src/chip.rs:185]:
```suggestion
                // As the current PIO interface does not provide support for interrupts, they are
```

**Review by alevy — APPROVED** [review]
